### ISSUE TITLE
PSD estimation is potentially a rate-limiting step in the search workflow

### ISSUE THREAD

**Issue by duncan-brown** [ligo-cbc/pycbc#377]
When running a search on more than a few wall-clock days of data, the PSD estimation jobs take an extremely long time and can be the rate limiting step in the search. See for example:

http://sugar-dev3.phy.syr.edu/pegasus/u/dbrown/r/66/w/2/j/30/ji/91

which has been running for over a day. The jobs should be parallelized across nodes in the search and PSD estimation workflows.

**Comment by ahnitz** [comment]
Duncan,

It doesn't say where the job is running, is that available to you? In any case I'll start the same job on spice-dev1 and see how long it takes.

**Comment by duncan-brown** [comment]
Take a look at this one:
```
-- Submitter: sugar-dev3.phy.syr.edu : <128.230.146.15:9619> : sugar-dev3.phy.syr.edu
 ID          OWNER            SUBMITTED     RUN_TIME HOST(S)
25597057.0   calculate_psd-  9/26 13:33   0+17:30:25 slot1@crush0733.sugar.phy.syr.edu
```
```
62652 dbrown    20   0  646m  90m 3948 R 89.0  0.2 897:52.49 pycbc_calculate                                                                                                                 
62659 dbrown    20   0  646m  87m 3948 R 85.7  0.2 895:45.34 pycbc_calculate                                                                                                                 
62657 dbrown    20   0  644m  86m 3948 R 84.4  0.2 896:53.67 pycbc_calculate                                                                                                                 
62660 dbrown    20   0  631m  75m 3948 R 83.7  0.2 936:16.71 pycbc_calculate                                                                                                                 
62658 dbrown    20   0  631m  73m 3948 R 83.1  0.2 942:36.24 pycbc_calculate                                                                                                                 
62655 dbrown    20   0  636m  79m 3948 R 80.5  0.2 920:32.04 pycbc_calculate                                                                                                                 
62654 dbrown    20   0  629m  70m 3948 R 71.6  0.2 944:26.94 pycbc_calculate                                                                                                                 
62656 dbrown    20   0  626m  69m 3948 D 52.2  0.2 933:05.92 pycbc_calculate                                                                                                                 
62653 dbrown    20   0  626m  67m 3948 D 50.9  0.2 926:57.77 pycbc_calculate                                                                                                                 
62661 dbrown    20   0  626m  67m 3948 D 43.3  0.2 925:25.95 pycbc_calculate                                                                                                                 
```

**Comment by duncan-brown** [comment]
I can't run this job from the command line:
```shell
(pycbc)[dbrown@sugar-dev3 ~]$ ./run.sh 
./run.sh: line 2: /home/dbrown/src/pycbc/bin/pycbc_calculate_psd: Argument list too long
(pycbc)[dbrown@sugar-dev3 ~]$ wc run.sh 
      2   78463 3216020 run.sh
```
The command line is over three million characters, because of all the frame files. I think PSD estimation definitely needs to be split into several jobs in the workflow and then joined back together at the end.

**Comment by ahnitz** [comment]
I was able to run on the command line (using internal datafind), and it finished in 3.5 hours, which I think is fine, considering that you seem to be running over a couple months of data. 

See /home/ahnitz/projects/psd_test/h.sh for an example. 

I'll see if I can replicate what is happening on the crush node.

**Comment by spxiwh** [comment]
Hi Alex, Duncan,

I'm not sure that this *needs* to be considered for 1.2.0, but I have also encountered this being the limiting factor in tuning workflows with small template banks (in these cases I tend to just re-use some old file with pegasus).

@ahnitz, is there any reason *not* to parallelise this as Duncan suggests? Is there some technical complication to that approach? If not, I would suggest to do this anyway.

**Comment by ahnitz** [comment]
Sorry, there is no technical reason not to. I just want to make sure I
understand why this happening as in my own tests this job is limited by
frame reading. Is it easier for the frame server if jobs are on different
nodes maybe?

**Comment by duncan-brown** [comment]
@ahnitz @spxiwh the reason I'd like to get this in 1.2.0 is that it's rate limiting for several workflows, and I don't want to have to hand-run the PSD estimation on head nodes and this seems to be a limiting step in running the workflows. I did notice that HyperThreading is enabled on the crush nodes. I will try limiting the number of running jobs to the number of physical processes and see if that helps.

But I agree with Ian that we should ultimately parallelize this anyway.

**Comment by duncan-brown** [comment]
OK, I've limited Condor to use the number of physical CPUs on crush and spice nodes. If the PSDs complete in a few hours, then we can leave this on the todo list, otherwise, I'll add the 1.2.0 milestone back.

**Comment by duncan-brown** [comment]
@ahnitz here is another instance running on a spice node:
```
25653770.0     |-calculate_psd 9/27 14:28   0+08:44:01 slot1@spice112.sugar.phy.syr.edu
25653771.0     |-calculate_psd 9/27 14:28   0+04:34:45 slot1@spice108.sugar.phy.syr.edu
```
the processes are in R state (one is in D), they haven't been evicted, and are not swapping:
```
top - 19:25:18 up  4:39,  1 user,  load average: 12.66, 12.82, 13.17
Tasks: 465 total,  13 running, 452 sleeping,   0 stopped,   0 zombie
Cpu(s): 29.7%us,  0.3%sy,  0.0%ni, 63.5%id,  6.4%wa,  0.0%hi,  0.0%si,  0.0%st
Mem:  49416280k total, 13310672k used, 36105608k free,    23172k buffers
Swap: 16252924k total,    26276k used, 16226648k free, 10489416k cached

  PID USER      PR  NI  VIRT  RES  SHR S %CPU %MEM    TIME+  COMMAND                                                                                                                    
 6984 dbrown    20   0  630m 100m 4936 R 99.7  0.2 114:23.56 pycbc_calculate                                                                                                             
 6986 dbrown    20   0  630m 100m 4936 R 99.7  0.2 113:41.08 pycbc_calculate                                                                                                             
 6988 dbrown    20   0  629m  98m 4936 R 99.7  0.2 114:41.32 pycbc_calculate                                                                                                             
26944 dbrown    20   0 1363m 672m  27m R 99.7  1.4  34:06.48 pycbc_inspiral                                                                                                              
31298 dbrown    20   0 1122m 612m  21m R 99.7  1.3   1:42.77 pycbc_inspiral                                                                                                              
 6987 dbrown    20   0  630m 100m 4936 R 95.8  0.2 114:57.99 pycbc_calculate                                                                                                             
 6985 dbrown    20   0  631m 100m 4936 R 93.8  0.2 115:06.29 pycbc_calculate                                                                                                             
 6989 dbrown    20   0  630m 100m 4936 R 93.8  0.2 114:57.09 pycbc_calculate                                                                                                             
 6990 dbrown    20   0  630m 100m 4936 R 93.8  0.2 114:38.38 pycbc_calculate                                                                                                             
 6992 dbrown    20   0  630m 100m 4936 R 91.9  0.2 113:38.63 pycbc_calculate                                                                                                             
 6993 dbrown    20   0  630m 100m 4936 D 91.9  0.2 113:34.21 pycbc_calculate                                                                                                             
 6991 dbrown    20   0  630m 100m 4936 R 82.1  0.2 113:56.17 pycbc_calculate                                                                                                             
```

**Comment by ahnitz** [comment]
Duncan,

Your own numbers say that there was a lot of system waiting. The jobs
themselves have only run for less than two hours in this report, when
they've been "running" from condors point of view for either 2 or 4x longer
(depending on which node you logged into). My guess is that you only
happened to catch them when there was a temporary reprieve in the frame
server response.

I tested this last night, and I saw a similar thing where at the beginning
they are mostly stuck in D state, but they did manage to complete. I think
this is much worse than when I did my initial testing, so the frame server
must be getting pounded and isn't really keeping up at all.

In any case, I did manage to implement a parallelization of this, but I
think it will still be mostly wasted cycles. Maybe this will help in the
short term though. Longer term, I would like a way to run jobs more
efficiently that require the nfs server where the frames are stored to be
more responsive. I am still wondering why the head nodes have such better
access to the frames? Are they on a different network?

-Alex
